core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala

```diff
@@ -77,12 +77,32 @@ private[spark] class HeartbeatReceiver(sc: SparkContext, clock: Clock)
 
   private[spark] var scheduler: TaskScheduler = null
 
-  // executor ID -> timestamp of when the last heartbeat from this executor was received
+  /**
+   * [SC-105641]
+   * Please make sure the intersection between `executorLastSeen` and `waitingList` is an empty set.
+   * If the intersection is not empty, it is possible to never kill the executor until the executor
+   * recovers. When an executor is in both `executorLastSeen` and `waitingList`, the value of
+   * `workerLastHeartbeat` in waitingList may update if the worker sends heartbeats to master
+   * normally.
+   *
+   * `executorLastSeen`:
+   *  - key: executor ID
+   *  - value: timestamp of when the last heartbeat from this executor was received
+   *
+   *  `waitingList`: executor ID -> WorkerLastHeartbeat
+   *  - key: executor ID
+   *  - value: timestamp of when the last heartbeat from the worker was received
+   *
+   * when driver does not receive any heartbeat from an executor for `executorTimeoutMs` seconds,
+   * the driver will ask master for the last heartbeat from the worker which the executor is running
+   * on.
+   */
   private val executorLastSeen = new HashMap[String, Long]
+  private val waitingList = new HashMap[String, Long]
```

Comment:
We need `waitingList` to be updated/maintained only when `HEARTBEAT_RECEIVER_CHECK_WORKER_LAST_HEARTBEAT` is enabled.

Reply:
Updated. 

Originally, I updated/maintained `waitingList` no matter the value of `HEARTBEAT_RECEIVER_CHECK_WORKER_LAST_HEARTBEAT` to make dynamic configuration implementation convenient. Currently, we only update/maintain `waitingList` when `checkWorkerLastHeartbeat` is true.